messaging/src/main/java/org/axonframework/commandhandling/gateway/CommandResult.java

```diff
@@ -16,47 +16,122 @@
 
 package org.axonframework.commandhandling.gateway;
 
+import jakarta.annotation.Nonnull;
 import org.axonframework.messaging.Message;
 
 import java.util.concurrent.CompletableFuture;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Interface describing the result from command handling.
+ *
+ * @author Allard Buijze
+ * @since 0.6.0
+ */
 public interface CommandResult {
 
-    default <R> CompletableFuture<R> resultAs(Class<R> type) {
+    /**
+     * Returns the result {@link Message} within a {@link CompletableFuture} of command handling.
+     *
+     * @return The result {@link Message} within a {@code CompletableFuture} of command handling.
+     */
+    CompletableFuture<? extends Message<?>> getResultMessage();
+
+    /**
+     * Returns the result of command handling, cast to the given {@code type}, within a {@link CompletableFuture}.
+     *
+     * @param type The expected result type of command handling.
+     * @param <R>  The type of the command result.
+     * @return The result of command handling, cast to the given {@code type}, within a {@code CompletableFuture}.
+     */
+    default <R> CompletableFuture<R> resultAs(@Nonnull Class<R> type) {
         return getResultMessage().thenApply(Message::getPayload)
-                                 .thenApply(type::cast);
+                                 .thenApply(payload -> requireNonNull(type, "The result type must not be null")
```

Comment:
I see several occasions in this file where we do `requireNonNull` inside lambdas and results.
I think it's better to do the validation up front. So:

```java
        requireNonNull(type, "The result type must not be null")
        return getResultMessage().thenApply(Message::getPayload)
                                 .thenApply(payload -> type.cast(payload))
```

This way we can fail-fast before wasting resources, and the lambda is easier to read :) 
Not that this method uses many many resources, but the `onSuccess` potentially does.

Reply:
Good point, will adjust!